tokio/src/io/util/buf_writer.rs

```diff
@@ -131,6 +131,67 @@ impl<W: AsyncWrite> AsyncWrite for BufWriter<W> {
         }
     }
 
+    fn poll_write_vectored(
+        mut self: Pin<&mut Self>,
+        cx: &mut Context<'_>,
+        bufs: &[IoSlice<'_>],
```

Comment:
In general, I think that this could be simplified quite a lot if you make use of the fact that it is possible to reassign `bufs` to a sub-slice of itself to remove slices from the beginning of the buffer.

Reply:
Thank you for the suggestion. I could not simplify by much, but hopefully the code is more readable now without an explicit iterator.

Reply:
I gave it a try myself and came up with this. What do you think?
```Rust
} else {
    // Remove empty buffers at the beginning of bufs.
    while bufs.first().map(|buf| buf.len()) == Some(0) {
        bufs = &bufs[1..];
    }
    if bufs.is_empty() {
        return Poll::Ready(Ok(0));
    }

    // Flush if the first buffer doesn't fit.
    let first_len = bufs[0].len();
    if first_len >= self.buf.capacity() - self.buf.len() {
        ready!(self.as_mut().flush_buf(cx))?;
        debug_assert!(self.buf.is_empty());
    }

    let me = self.project();

    if first_len >= me.buf.capacity() {
        // The slice is at least as large as the buffering capacity,
        // so it's better to write it directly, bypassing the buffer.
        return me.inner.poll_write(cx, &*bufs[0]);
    }

    // Append the buffers that fit in the internal buffer.
    // The first buffer is guaranteed to fit due to the flush above.
    let mut total_written = 0;
    for buf in bufs {
        if buf.len() > me.buf.capacity() - me.buf.len() {
            break;
        } else {
            me.buf.extend_from_slice(&*buf);
            total_written += buf.len();
        }
    }

    debug_assert!(total_written != 0);
    Poll::Ready(Ok(total_written))
}
```

Reply:
@Darksonn Thanks! I have unrolled the first iteration of appending though, since we're into improving performance and the check in the loop body would be done redundantly, and there is a suitable `else` branch for that.